### PR TITLE
test: e2e dispute

### PR DESCRIPTION
**Pull request by 0xyaco**
# 🤖 Linear

Closes GRT-192, GRT-216

## Description
* Test agent reaction to wrong response

Miscelanous fixes:
* Improve `waitForEvent` E2E testing util to return the found event
* Move dispute status mapping into `ProphetCodec` class
* Normalize all IDs read from events as soon as they are fetched
* Fix dispute settling errors not being handled correctly (thus not escalating if needed)
* Fix proposing a new response when the active response has been disputed

### CODE REVIEW

File: apps/agent/test/e2e/scenarios/01_happy_path/index.spec.ts [hunk 1]
```diff
@@ -213,30 +221,367 @@ describe.sequential("single agent", () => {
...
+     *      - `ResponseDisputed(RESP1.id, DISP1.id, DISP1)`
+     *  - A1 proposes RESP2
+     *      - `ResponseProposed(REQ1.id, RESP2.id, RESP2)`
+     *  - A1 settles DISP1 and it ends with status `Won`
+     *      - `DisputeStatusUpdated(DISP1.id, DISP1, "Won")`
+     *  - A1 finalizes REQ1 with RESP2
+     *      - `EBOFinalityModule.newEpoch(E1, CHAIN1, RESP2.response)`
+     *      - `OracleRequestFinalized(REQ1.id, RESP2.id, A1.address)`
+     */
+    test.skip("dispute response and propose a new one", { timeout: E2E_TEST_TIMEOUT }, async () => {
```
[the remark this answers is not in the document]
Comment (0xyaco) [reply apps/agent/test/e2e/scenarios/01_happy_path/index.spec.ts:352]: Yep, the idea is to enable every test to be run sequentially later

File: packages/automated-dispute/src/services/eboActor.ts [hunk 2]
```diff
@@ -674,9 +727,10 @@ export class EboActor {
         const dispute: Dispute["prophetData"] = {
             disputer: disputer,
             proposer: proposedResponse.prophetData.proposer,
-            responseId: HexUtils.normalize(event.metadata.responseId) as ResponseId,
+            responseId: event.metadata.responseId,
```
Comment (0xnigir1) [inline packages/automated-dispute/src/services/eboActor.ts:730]: just double check that normalizing is no longer required

File: packages/automated-dispute/src/services/eboActor.ts [hunk 3]
```diff
@@ -761,8 +815,32 @@ export class EboActor {
...
+
+                            throw err;
+                        }
+                    }
+                })(),
+            ]);
+
+            operations.forEach((element) => {
+                if (element.status === "rejected") throw element.reason;
+            });
```
Comment (0xnigir1) [inline packages/automated-dispute/src/services/eboActor.ts:842]: if we're throwing on 'rejected', there's no benefit on doing `Promise.allSettled` unless we require that Promises are all fulfilled

File: packages/shared/src/services/stringify.ts
Comment (0xnigir1) [inline packages/shared/src/services/stringify.ts:1]: random fact: Viem has this stringify implemented

File: packages/automated-dispute/src/services/eboActor.ts [hunk 4]
```diff
@@ -343,6 +333,44 @@ export class EboActor {
...
+
+    private async finalizeRequest(request: Request, response: Response) {
+        this.logger.info(`Finalizing request.`);
+        this.logger.debug(stringify({ request: request, response: response }));
+
+        try {
+            await this.protocolProvider.finalize(request.prophetData, response.prophetData);
+        } catch (err) {
+            if (err instanceof CustomContractError) {
+                err.setContext({
```
[the remark this answers is not in the document]
Comment (0xyaco) [reply packages/automated-dispute/src/services/eboActor.ts:377]: Added a warn here too!

**Review by 0xnigir1 — COMMENTED** [review]
wonderful, few comments but nothing critical. after addressing beebs comments is good to go to me

**Pull request merged**